File: backend/app/api/v1/cameras.py
"""``/cameras/*`` — the server-side camera registry (2026-09-02).

* ``GET    /cameras``                 — Page[CameraRead]
* ``GET    /cameras/all``             — every camera, unpaged (the globe wants all of them)
* ``POST   /cameras``                 — 201 CameraRead
* ``POST   /cameras/import``          — 201 CameraImportResult (browser registry / JSON)
* ``GET    /cameras/{id}``            — CameraRead
* ``GET    /cameras/{id}/export``     — the whole camera as one zip: the registry row,
                                      the frame, the control points, the DEM, the DEM
                                      calibration, the lookup table and the drift reference
* ``PATCH  /cameras/{id}``            — CameraRead
* ``PUT    /cameras/{id}/desired``    — CameraRead (what the camera SHOULD be doing)
* ``POST   /cameras/{id}/drift/freeze`` — CameraDriftRead: freeze the drift reference on
                                      the camera's frame and watch it in the background
* ``DELETE /cameras/{id}``            — 204; stops the camera's watch / run / feed first

★ ``desired`` is normally written by the detection, drift and live routers when a
start or stop carries a ``camera_id`` — the explicit PUT exists for the page's
"forget this setup" and for scripts. Read ``services/camera_service.py`` for the
intent-in / status-out rule before adding anything here that reads a thread.

★ ``/cameras/all`` and ``/cameras/import`` are registered BEFORE ``/cameras/{id}`` so
a literal segment is never parsed as a UUID (``assert_no_duplicate_routes`` proves
the two literals do not collide with anything).
"""

from __future__ import annotations

from pathlib import Path
from uuid import UUID

from fastapi import (
    APIRouter,
    BackgroundTasks,
    Depends,
    HTTPException,
    Query,
    Request,
    Response,
    status,
)
from fastapi.concurrency import run_in_threadpool
from fastapi.responses import FileResponse
from sqlalchemy.ext.asyncio import AsyncSession

from app.api import deps
from app.core.config import Settings
from app.core.exceptions import EmptyPatch
from app.core.pagination import SortKey, SortParams
from app.schemas.camera import (
    CAMERA_SORT_FIELDS,
    CameraCreate,
    CameraDesired,
    CameraDriftRead,
    CameraImportRequest,
    CameraImportResult,
    CameraRead,
    CameraUpdate,
    desired_from_row,
)
from app.schemas.common import Page
from app.services import camera_service, drift_service
from app.services.camera_service import CameraService

router = APIRouter(prefix="/cameras", tags=["cameras"])

_DEFAULT_SORT = (SortKey("name", descending=False),)


def to_camera_read(camera) -> CameraRead:  # noqa: ANN001 — ORM row; the module's one presenter
    return CameraRead(
        id=camera.id,
        name=camera.name,
        lat=camera.lat,
        lon=camera.lon,
        source=camera.source,
        connection=camera.connection,
        provides=camera.provides,
        data_source=camera.data_source,
        heading_deg=camera.heading_deg,
        fov_deg=camera.fov_deg,
        fps=camera.fps,
        tags=list(camera.tags or []),
        lut_site=camera.lut_site,
        project_id=camera.project_id,
        frame_image_id=camera.frame_image_id,
        calibration=camera.calibration,
        desired=desired_from_row(camera.desired),
        created_at=camera.created_at,
        updated_at=camera.updated_at,
    )


@router.get("", response_model=Page[CameraRead], summary="List cameras")
async def list_cameras(
    pagination=Depends(deps.get_pagination),
    sort: str | None = Depends(deps.get_sort),
    q: str | None = Query(None, description="Free-text over name and sources."),
    db: AsyncSession = Depends(deps.get_db),
) -> Page[CameraRead]:
    parsed = SortParams.parse(sort, allowed=CAMERA_SORT_FIELDS, default=_DEFAULT_SORT)
    rows, total = await CameraService(db).list(pagination=pagination, sort=parsed, q=q)
    return Page.of([to_camera_read(c) for c in rows], total, pagination)


@router.get("/all", response_model=list[CameraRead], summary="Every camera, by name")
async def list_all_cameras(db: AsyncSession = Depends(deps.get_db)) -> list[CameraRead]:
    """★ The one unpaged list: the globe draws EVERY camera at once and a site has
    tens of them, not thousands. A registry that grows past that gets a real page."""
    return [to_camera_read(c) for c in await CameraService(db).all()]


@router.post(
    "",
    response_model=CameraRead,
    status_code=status.HTTP_201_CREATED,
    summary="Register a camera",
)
async def create_camera(
    body: CameraCreate,
    request: Request,
    response: Response,
    db: AsyncSession = Depends(deps.get_db),
    _: None = Depends(deps.require_writable),
) -> CameraRead:
    camera = await CameraService(db).create(body)
    response.headers["Location"] = f"{request.scope.get('root_path', '')}/api/v1/cameras/{camera.id}"
    return to_camera_read(camera)


@router.post(
    "/import",
    response_model=CameraImportResult,
    status_code=status.HTTP_201_CREATED,
    summary="Import cameras (a browser registry, or a JSON export)",
)
async def import_cameras(
    body: CameraImportRequest,
    db: AsyncSession = Depends(deps.get_db),
    _: None = Depends(deps.require_writable),
) -> CameraImportResult:
    """★ All or nothing, and the result maps every ``client_id`` to its server UUID
    so the browser can re-key its per-camera settings once, on migration."""
    return await CameraService(db).import_many(body)


@router.get("/{camera_id}", response_model=CameraRead, summary="Get a camera")
async def get_camera(camera_id: UUID, db: AsyncSession = Depends(deps.get_db)) -> CameraRead:
    return to_camera_read(await CameraService(db).get(camera_id))


@router.patch("/{camera_id}", response_model=CameraRead, summary="Update a camera")
async def update_camera(
    camera_id: UUID,
    body: CameraUpdate,
    db: AsyncSession = Depends(deps.get_db),
    _: None = Depends(deps.require_writable),
) -> CameraRead:
    if body.is_empty():
        raise EmptyPatch("The PATCH body is empty; send at least one field to change.")
    return to_camera_read(await CameraService(db).update(camera_id, body))


@router.put(
    "/{camera_id}/desired",
    response_model=CameraRead,
    summary="Set what this camera should be doing (restored at boot)",
)
async def set_camera_desired(
    camera_id: UUID,
    body: CameraDesired,
    db: AsyncSession = Depends(deps.get_db),
    _: None = Depends(deps.require_writable),
) -> CameraRead:
    """★ Intent only. Setting ``watch`` here does not start a watch NOW — the drift
    router does that (and records intent as a side effect). This is the explicit
    door: clear a stale intent, or pre-register one for the next boot."""
    return to_camera_read(await CameraService(db).set_desired(camera_id, body))


@router.post(
    "/{camera_id}/drift/freeze",
    response_model=CameraDriftRead,
    status_code=status.HTTP_201_CREATED,
    summary="Freeze the drift reference on the camera's frame and watch it in the background",
)
async def freeze_camera_drift(
    camera_id: UUID,
    db: AsyncSession = Depends(deps.get_db),
    settings: Settings = Depends(deps.get_settings),
    _: None = Depends(deps.require_writable),
) -> CameraDriftRead:
    """★ THE FRAME WITH THE CONTROL POINTS IS THE FROZEN FRAME (2026-09-08). The
    camera settings call this when the lookup table is built and whenever a new
    frame is chosen; the monitoring page only reads the verdict. The watch it
    starts is recorded as the camera's desired state, so it survives a restart.
    """
    from app.api.v1.drift import resolve_image_frame  # noqa: PLC0415 — sibling router's resolver

    service = CameraService(db)
    camera = await service.get(camera_id)
    image_path = image_label = None
    if camera.frame_image_id is not None:
        image_path, image_label = await resolve_image_frame(camera.frame_image_id, db, settings)
    previous = desired_from_row(camera.desired).watch
    try:
        outcome = await run_in_threadpool(
            camera_service.refreeze_drift_watch,
            settings,
            camera_id=str(camera.id),
            name=camera.name,
            source=camera.source or "",
            lut_site=camera.lut_site,
            image_path=image_path,
            image_label=image_label,
            fov_deg=camera.fov_deg,
            previous_ref_id=previous.ref_id if previous is not None else None,
        )
    except drift_service.DriftError as exc:
        raise HTTPException(
            status_code=status.HTTP_422_UNPROCESSABLE_ENTITY, detail=str(exc)
        ) from exc
    await service.record_watch(
        camera.id, {"ref_id": outcome["ref_id"], "interval_s": outcome["interval_s"]}
    )
    return CameraDriftRead(**outcome)


@router.get(
    "/{camera_id}/export",
    summary="The whole camera as one zip — settings, frame, GCPs, DEM, calibration, LUT, drift",
)
async def export_camera(
    camera_id: UUID,
    background: BackgroundTasks,
    settings: Settings = Depends(deps.get_settings),
) -> Response:
    """Everything this camera IS, in one downloadable folder.

    ★ A PARTIAL CAMERA STILL EXPORTS. Setup is eight steps and an operator may be
      on step three; every step not done yet is named in the bundle's
      ``camera.json`` under ``missing`` rather than refusing the download. What
      comes back is always a true account of how far this camera got.
    """
    from app.services import camera_export_service

    try:
        path, filename = await run_in_threadpool(
            camera_export_service.export_camera, str(camera_id), settings
        )
    except camera_export_service.CameraExportError as exc:
        raise HTTPException(status_code=status.HTTP_404_NOT_FOUND, detail=str(exc)) from exc
    # The zip is a temporary file; it goes once the response has been sent.
    background.add_task(lambda: Path(path).unlink(missing_ok=True))
    return FileResponse(
        path,
        media_type="application/zip",
        filename=filename,
        background=background,
        headers={"Cache-Control": "no-store"},
    )


@router.delete(
    "/{camera_id}",
    status_code=status.HTTP_204_NO_CONTENT,
    summary="Remove a camera (its detection records stay)",
)
async def delete_camera(
    camera_id: UUID,
    db: AsyncSession = Depends(deps.get_db),
    _: None = Depends(deps.require_writable),
) -> Response:
    await CameraService(db).delete(camera_id)
    return Response(status_code=status.HTTP_204_NO_CONTENT)

File: frontend/src/api/cameras.ts
/**
 * Cameras — the server-side registry (`/cameras/*`, 2026-09-02).
 *
 * ★ THE REGISTRY MOVED. Until 1.3 the cameras lived in ONE browser's
 *   `localStorage` (`le.cameras.v1`); a second machine saw nothing, and the
 *   `detection_events` a camera wrote named it by a string only that browser knew.
 *   `cameraRegistryStore` is now a CACHE of this API: every add/patch/remove goes
 *   through here first, and the store takes the server's answer.
 *
 * ★ `desired` IS INTENT, NOT STATUS. `CameraDesired` is what the operator asked the
 *   camera to be doing (a watch, a run, a feed); the API restarts it after a
 *   restart. Whether it is doing it NOW is the detection / drift / feed status.
 *
 * ★ No `fetch` here: every call goes through `client.ts` (§2.5).
 */

/* eslint-disable camelcase -- wire fields are snake_case (L9) */
import type { Page, Uuid } from '../types/common';
import { API_BASE_URL, fetchJson, toQuery } from './client';

/** ★ The three things that differ (narrowed 2026-09-11, see `cameraRegistryStore`):
 *  an address on the network, a capture device on this machine, a serial line. */
export type CameraConnection = 'lan' | 'usb' | 'serial';

/** The optional calibration entered on the camera itself — `ImageCameraPut`'s
 *  vocabulary minus the position (the registry's lat/lon IS the position). */
export interface CameraCalibration {
  fx: number | null;
  fy: number | null;
  cx: number | null;
  cy: number | null;
  k1: number | null;
  k2: number | null;
  p1: number | null;
  p2: number | null;
  k3: number | null;
  mast_offset_m: number | null;
  tilt_deg: number | null;
}
export type CameraProvides = 'camera' | 'data' | 'both';

export interface CameraDesiredWatch {
  ref_id: string;
  interval_s: number;
}

export interface CameraDesiredDetect {
  lut_site: string | null;
  conf: number;
  classes: number[] | null;
  imgsz: number;
  tile: boolean;
  model: string | null;
  tracker_start_frame: number;
  tracker_type: string;
  tracker_refresh_every: number;
  centre_marks: boolean;
  /** Steady boxes — absent on intents recorded before 2026-09-03 (defaults on). */
  steady_boxes?: boolean;
}

export interface CameraDesired {
  watch: CameraDesiredWatch | null;
  detect: CameraDesiredDetect | null;
  feed: boolean;
}

/** The registry attributes — what the add dialog collects, what the row stores. */
export interface CameraFields {
  name: string;
  lat: number;
  lon: number;
  source: string | null;
  connection: CameraConnection;
  provides: CameraProvides;
  data_source: string | null;
  heading_deg: number | null;
  fov_deg: number | null;
  fps: number | null;
  tags: string[];
  lut_site: string | null;
  /** The camera's backing project — its DEM, frame and control points live there. */
  project_id: Uuid | null;
  /** The frame chosen from this camera — the photograph the control points sit on. */
  frame_image_id: Uuid | null;
  calibration: CameraCalibration | null;
}

export interface CameraRead extends CameraFields {
  id: Uuid;
  desired: CameraDesired;
  created_at: string;
  updated_at: string;
}

/** `POST /cameras` body — every field optional except name/lat/lon. */
export type CameraCreate = Partial<CameraFields> & Pick<CameraFields, 'name' | 'lat' | 'lon'>;
/** `PATCH /cameras/{id}` — only the sent keys change. */
export type CameraUpdate = Partial<CameraFields>;

export interface CameraImportItem extends CameraCreate {
  /** The browser's old id (a ULID) — the result maps it to the server UUID. */
  client_id?: string;
}

export interface CameraImportResult {
  created: number;
  items: { client_id: string | null; id: Uuid; name: string }[];
}

/**
 * `POST /cameras/{id}/drift/freeze` — the drift reference now frozen on the
 * camera's FRAME (the photograph its control points sit on), and the background
 * watch on it (2026-09-08, owner decision). The monitoring page only reads.
 */
export interface CameraDriftRead {
  ref_id: string;
  interval_s: number;
  n_landmarks: number;
  frozen_from_label: string | null;
  created_utc: string;
  intrinsics_mode: 'calibration' | 'fov';
  intrinsics_warning: string | null;
  watching: boolean;
}

const path = (id: Uuid): string => `/cameras/${id}`;

export const camerasApi = {
  /** `GET /cameras` → `Page[CameraRead]`. */
  list: (q?: string, signal?: AbortSignal): Promise<Page<CameraRead>> =>
    fetchJson<Page<CameraRead>>('/cameras', { query: toQuery({ q, limit: 200 }), signal }),

  /** `GET /cameras/all` → every camera, by name — the globe draws all of them. */
  all: (signal?: AbortSignal): Promise<CameraRead[]> =>
    fetchJson<CameraRead[]>('/cameras/all', { signal }),

  /** `GET /cameras/{id}`. */
  get: (id: Uuid, signal?: AbortSignal): Promise<CameraRead> =>
    fetchJson<CameraRead>(path(id), { signal }),

  /**
   * ★ THE WHOLE CAMERA, ONE FILE (2026-09-12, owner ask). A zip holding the
   * registry row — name, connection, position, tags, calibration — beside every
   * artefact its setup produced: the frame captured and used, the control points,
   * the DEM, the processed DEM calibration, the lookup table and the frozen drift
   * reference. A half-configured camera exports too; the bundle's `camera.json`
   * names each step that had not happened rather than refusing.
   *
   * A direct URL, not a `fetchJson`: the browser downloads it.
   */
  exportUrl: (id: Uuid): string => `${API_BASE_URL}${path(id)}/export`,

  /** `POST /cameras` → `201 CameraRead`. A 422 names the rule that refused it. */
  create: (body: CameraCreate, signal?: AbortSignal): Promise<CameraRead> =>
    fetchJson<CameraRead>('/cameras', { method: 'POST', body, signal }),

  /** `PATCH /cameras/{id}` — send ONLY the changed fields. */
  update: (id: Uuid, body: CameraUpdate, signal?: AbortSignal): Promise<CameraRead> =>
    fetchJson<CameraRead>(path(id), { method: 'PATCH', body, signal }),

  /** `DELETE /cameras/{id}` → 204. Its watch / run / feed are stopped first. */
  remove: async (id: Uuid, signal?: AbortSignal): Promise<void> => {
    await fetchJson<void>(path(id), { method: 'DELETE', parse: 'none', signal });
  },

  /** `POST /cameras/import` → `201`. All or nothing; maps client ids to UUIDs. */
  import: (cameras: CameraImportItem[], signal?: AbortSignal): Promise<CameraImportResult> =>
    fetchJson<CameraImportResult>('/cameras/import', { method: 'POST', body: { cameras }, signal }),

  /** `PUT /cameras/{id}/desired` — intent only; starts nothing NOW. */
  setDesired: (id: Uuid, desired: CameraDesired, signal?: AbortSignal): Promise<CameraRead> =>
    fetchJson<CameraRead>(`${path(id)}/desired`, { method: 'PUT', body: desired, signal }),

  /**
   * `POST /cameras/{id}/drift/freeze` → 201. Freezes the drift reference on the
   * camera's frame with its lookup table and starts the background watch; a 422
   * names the missing piece (the table, the frame, or the pose).
   */
  freezeDrift: (id: Uuid, signal?: AbortSignal): Promise<CameraDriftRead> =>
    fetchJson<CameraDriftRead>(`${path(id)}/drift/freeze`, { method: 'POST', signal }),
};
